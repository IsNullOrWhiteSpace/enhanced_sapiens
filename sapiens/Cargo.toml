[package]
name = "sapiens"
version = "0.10.2"
edition.workspace = true
authors.workspace = true
description = "Core - Sapiens"
repository.workspace = true
license.workspace = true
publish.workspace = true
readme = "../README.md"
keywords.workspace = true
categories.workspace = true
documentation = "https://docs.rs/sapiens"

[features]
default = ["clap"]

# derive Clap traits
clap = ["dep:clap"]

[dependencies]
tokio = { version = "1.34.0" }
tracing = "0.1.40"
async-trait = "0.1.74"
lazy_static = "1.4.0"

serde = { version = "1.0.193", features = ["derive"] }
serde_yaml = "0.9.27"

clap = { version = "4.5.1", optional = true }

# OpenAI API - OpenAI and lm-sys/FastChat
async-openai = "0.16.3"
tokenizers = { version = "0.15.0", features = [] }

# GCP Vertex AI Generative Language Models
gcp-vertex-ai-generative-language = "0.1.2"

thiserror = "1.0.50"

[dev-dependencies]
indoc = "2"
insta = { version = "1.34.0", features = ["yaml"] }
