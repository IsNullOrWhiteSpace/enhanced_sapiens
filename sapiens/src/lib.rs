//! Sapiens library
//!
//! *Sapiens uses tools to interact with the world.*
//!
//! An experiment with handing over the tools to the machine.
//!
//! # Overview
//! This library is the core of Sapiens. It contains the logic for the
//! interaction between the user, the language model and the tools.
//!
//! # More information
//! See https://github.com/ssoudan/sapiens/tree/main/sapiens_cli for an example of usage or
//! https://github.com/ssoudan/sapiens/tree/main/sapiens_bot for a Discord bot.
//!
//! https://github.com/ssoudan/sapiens/tree/main/sapiens_exp is a framework to run experiments and collect traces
//! of the interactions between the language model and the tools to accomplish a
//! task.
//!
//! A collection of tools is defined in https://github.com/ssoudan/sapiens/tree/main/sapiens_tools.
pub mod context;

/// Prompt generation logic
pub mod prompt;

/// Toolbox for sapiens
pub mod tools;

/// Language models
pub mod models;

pub mod chains;

use std::fmt::Debug;
use std::str::FromStr;
use std::sync::{Arc, Weak};

use clap::builder::PossibleValue;
use serde::{Deserialize, Serialize};
use tokio::sync::Mu