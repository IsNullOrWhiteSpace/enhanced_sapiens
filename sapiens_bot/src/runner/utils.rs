
use pulldown_cmark::{CodeBlockKind, Event, Options, Parser, Tag};
use pulldown_cmark_to_cmark::{cmark_resume, State};
use sapiens::chains::Message;
use sapiens::context::{ChatEntry, ChatEntryFormatter, MessageFormatter};
use sapiens::models::Role;

/// Chat entry formatter that renders the chat entry in markdown
pub struct Formatter {}

impl ChatEntryFormatter for Formatter {
    fn format(&self, entry: &ChatEntry) -> String {
        let msg = entry.msg.clone();
        match entry.role {
            Role::User => format!(":earth_americas:\n{}", msg),
            Role::Assistant => format!(":robot:\n{}", msg),
            Role::System => format!(":rooster:\n{}", msg),
            Role::Function => format!(":gear:\n{}", msg),
            Role::Tool => format!(":wrench:\n{}", msg),
        }
    }
}

impl MessageFormatter for Formatter {
    fn format(&self, msg: &Message) -> String {
        msg.to_string()
    }
}

/// Size in characters of an event once rendered in markdown
fn md_event_size(event: &Event) -> usize {
    match event {
        Event::Text(text) => text.len(),
        Event::Code(text) => text.len(),
        Event::Html(text) => text.len(),
        Event::FootnoteReference(text) => text.len(),
        Event::SoftBreak => 1,
        Event::HardBreak => 2,
        Event::Rule => 1,
        Event::TaskListMarker(_) => 3,
        Event::Start(tag) => {
            let len = match tag {
                Tag::Paragraph => 2,
                Tag::Heading(level, id, classes) => {
                    *level as usize
                        + 1
                        + id.map(|x| x.len()).unwrap_or_default()
                        + classes.iter().map(|x| x.len()).sum::<usize>()
                }
                Tag::BlockQuote => 2,
                Tag::CodeBlock(CodeBlockKind::Indented) => 2,
                Tag::CodeBlock(CodeBlockKind::Fenced(fence)) => fence.len() + 3,
                Tag::List(_) => 2,
                Tag::Item => 2,
                Tag::FootnoteDefinition(d) => d.len() + 3,
                Tag::Table(_) => 4,
                Tag::TableHead => 2,
                Tag::TableRow => 3,
                Tag::TableCell => 3,
                Tag::Emphasis => 2,
                Tag::Strong => 2,
                Tag::Strikethrough => 2,
                Tag::Link(_, u, t) => 4 + u.len() + t.len(),
                Tag::Image(_, u, t) => 4 + u.len() + t.len(),
            };
            len
        }
        Event::End(_tag) => 4, // random
    }
}

/// Sanitize messages for Discord
///
/// Tries to split mardown messages into multiple messages if they are too
/// long for Discord.
///
/// If any split is still too long, truncate it.
pub(crate) fn sanitize_msgs_for_discord(msgs: Vec<String>) -> Vec<String> {
    msgs.into_iter()
        .flat_map(|m| split_msgs(m, 1800))
        .map(|mut x| {
            if x.len() > 1800 - 3 {
                x.truncate(1800);
                x.push_str("...");
            }
            x
        })
        .collect()
}

fn is_block_delimiter(t: &Tag) -> bool {
    match t {
        Tag::Paragraph => true,
        Tag::Heading(_, _, _) => true,
        Tag::BlockQuote => true,
        Tag::CodeBlock(_) => true,
        Tag::List(_) => true,
        Tag::Item => false,