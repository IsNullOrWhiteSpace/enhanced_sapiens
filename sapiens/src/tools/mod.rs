
use std::collections::HashMap;
use std::fmt::Debug;

use serde::ser::SerializeMap;
use serde::{Deserialize, Serialize, Serializer};
use toolbox::Toolbox;
use tracing::warn;

use crate::tools::invocation::{ExtractedInvocations, InvocationError};

/// Tools to extract Tool invocations from a messages
pub mod invocation;

/// Collection of tools
pub mod toolbox;

/// Part of a [`Format`]
#[derive(Debug, Clone)]
pub struct FieldFormat {
    /// Name of the field
    pub name: String,
    /// Type of the field
    pub r#type: String,
    /// True if the field is optional
    pub optional: bool,
    /// Description of the field
    pub description: String,
}

/// Input or output format of a tool
pub trait Describe {
    /// Describe the format
    fn describe() -> Format;
}

/// Format of [`Tools`] input and output
#[derive(Debug, Clone, Default)]
pub struct Format {
    /// Fields of the format
    pub fields: Vec<FieldFormat>,
}

impl Serialize for Format {
    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let n = self.fields.len();
        let mut map = serializer.serialize_map(Some(n))?;
        for field in &self.fields {
            let description = if field.optional {
                format!("<{}> {} (optional)", field.r#type, field.description)
            } else {
                format!("<{}> {}", field.r#type, field.description)
            };
            map.serialize_entry(&field.name, &description)?;
        }
        map.end()
    }
}

impl From<Vec<FieldFormat>> for Format {
    fn from(fields: Vec<FieldFormat>) -> Self {
        Format { fields }
    }
}

/// Tool description
#[derive(Debug, Serialize, Clone)]
pub struct ToolDescription {
    /// Name of the tool
    pub name: String,
    /// Description of the tool
    pub description: String,
    /// Input format
    pub parameters: Format,
    /// Output format
    pub responses_content: Format,
}

impl ToolDescription {
    /// Create a new tool description
    pub fn new(
        name: &str,
        description: &str,
        parameters: Format,
        responses_content: Format,
    ) -> Self {
        ToolDescription {
            name: name.to_string(),
            description: description.to_string(),
            parameters,
            responses_content,
        }
    }
}

/// Error while using a tool
#[derive(Debug, thiserror::Error, Clone, Serialize, Deserialize)]
pub enum ToolUseError {
    /// Tool not found
    #[error("Tool not found: {0}")]
    ToolNotFound(String),
    /// Tool invocation failed
    #[error("Tool invocation failed: {0}")]
    InvocationFailed(String),
    /// Failed to serialize the output
    #[error("Failed to serialize the output: {0}")]
    InvalidOutput(String),
    /// Failed to deserialize the input
    #[error("Failed to deserialize the parameters: {0}")]
    InvalidInput(String),
}

/// A tool invocation input
#[derive(Serialize, Deserialize, Debug)]
pub(crate) struct ToolInvocationInput {
    /// The tool to invoke
    tool_name: String,
    // FUTURE(ssoudan) should this be flattened?
    // FUTURE(ssoudan) should this be called `spec` or `arguments` or `parameters`?
    /// The input to the tool
    parameters: serde_yaml::Value,
    /// The junk
    #[serde(skip_serializing_if = "HashMap::is_empty", flatten)]
    junk: HashMap<String, serde_yaml::Value>,
}

/// Something meant to become a [`Tool`] - description
pub trait ProtoToolDescribe {
    /// the description of the tool
    fn description(&self) -> ToolDescription;
}

/// Something meant to become a [`Tool`] - invocation
#[async_trait::async_trait]
pub trait ProtoToolInvoke {
    /// Invoke the tool
    async fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError>;
}

/// A Tool - the most basic kind of tools. See [`AdvancedTool`] and
/// [`TerminalTool`] for more.
#[async_trait::async_trait]
pub trait Tool: Sync + Send {
    /// the description of the tool
    fn description(&self) -> ToolDescription;

    /// Invoke the tool
    // FUTURE(ssoudan) Box<Deserialize>?
    async fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError>;
}

#[async_trait::async_trait]
impl<T: Sync + Send> Tool for T
where
    T: ProtoToolDescribe + ProtoToolInvoke,
{
    fn description(&self) -> ToolDescription {
        self.description()
    }

    async fn invoke(&self, input: serde_yaml::Value) -> Result<serde_yaml::Value, ToolUseError> {
        self.invoke(input).await
    }
}

/// A termination message
///
/// This is the message that is sent to the user when a chain of exchanges
/// terminates.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TerminationMessage {
    /// The final textual answer for this task.
    pub conclusion: String,
    /// The original question that was asked to the user.
    pub original_question: String,
}

/// A [`Tool`] that wraps a chain of exchanges
#[async_trait::async_trait]
pub trait TerminalTool: Tool + Sync + Send {
    /// done flag.
    async fn is_done(&self) -> bool {
        false
    }

    /// Take the done flag.
    async fn take_done(&self) -> Option<TerminationMessage> {
        None
    }
}

/// A [`Tool`]  that can benefit from a [`Toolbox`]
#[async_trait::async_trait]
pub trait AdvancedTool: Tool {
    /// Invoke the tool with a [`Toolbox`]
    async fn invoke_with_toolbox(
        &self,
        toolbox: Toolbox,
        input: serde_yaml::Value,
    ) -> Result<serde_yaml::Value, ToolUseError>;
}

async fn choose_invocation(
    tool_invocations: ExtractedInvocations,
) -> Result<ToolInvocationInput, InvocationError> {
    // TODO(ssoudan) customizable level of strictness
    if tool_invocations.yaml_block_count > 1 {
        return Err(InvocationError::TooManyYamlBlocks(
            tool_invocations.yaml_block_count,
        ));
    }

    // if no tool_invocations are found, we return an error
    if tool_invocations.invocations.is_empty() {
        return Err(InvocationError::NoInvocationFound);
    }

    // We just take the first one
    let mut invocation = tool_invocations.invocations.into_iter().next().unwrap();

    // FUTURE(ssoudan) clean up the object and return this one instead
    // if any tool_invocations have an 'output' field, we return an error
    if !invocation.junk.is_empty() {
        let junk_keys = invocation
            .junk
            .keys()
            .cloned()
            .collect::<Vec<String>>()
            .join(", ");

        warn!(
            ?junk_keys,
            "The Action should not have fields: {}.", junk_keys
        );

        // We just remove them for now
        invocation.junk.clear();

        // return Err(InvocationError::NoValidInvocationFound(format!(
        //     "The Action cannot have fields: {}. Only `command` and `input`
        // are allowed.",     junk_keys
        // )));
    }

    Ok(invocation)
}

#[cfg(test)]
mod tests {
    use std::collections::HashMap;

    use insta::assert_display_snapshot;
    use serde::{Deserialize, Serialize};

    #[derive(Debug, Serialize, Deserialize)]
    struct FakeToolInput {
        q: String,
        excluded_terms: Option<String>,
        num_results: Option<u32>,
    }

    #[derive(Debug, Serialize, Deserialize)]
    struct FakeToolOutput {
        items: Vec<String>,
    }

    #[tokio::test]
    async fn test_serializing_tool_invocation() {
        let input = FakeToolInput {
            q: "Marcel Deneuve".to_string(),
            excluded_terms: Some("Resident Evil".to_string()),
            num_results: Some(10),
        };

        let output = FakeToolOutput {
        items: vec![
            "Marcel Deneuve is a character in the Resident Evil film series,".to_string(), 
            "playing a minor role in Resident Evil: Apocalypse and a much larger".to_string(),
            " role in Resident Evil: Extinction. Explore historical records and ".to_string(),
            "family tree profiles about Marcel Deneuve on MyHeritage, the world's largest family network.".to_string()
        ]

        };

        let junk = vec![("output".to_string(), serde_yaml::to_value(output).unwrap())];

        let invocation = super::ToolInvocationInput {
            tool_name: "Search".to_string(),
            parameters: serde_yaml::to_value(input).unwrap(),
            junk: HashMap::from_iter(junk.into_iter()),
        };

        let serialized = serde_yaml::to_string(&invocation).unwrap();

        assert_display_snapshot!(serialized);
    }
}