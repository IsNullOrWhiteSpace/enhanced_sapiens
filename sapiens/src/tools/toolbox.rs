
use std::collections::HashMap;
use std::fmt::Debug;
use std::sync::Arc;

use serde::{Deserialize, Serialize};
use tokio::sync::RwLock;
use tracing::{debug, info};

use crate::tools;
use crate::tools::invocation::InvocationError;
use crate::tools::{
    AdvancedTool, TerminalTool, TerminationMessage, Tool, ToolDescription, ToolUseError,
};