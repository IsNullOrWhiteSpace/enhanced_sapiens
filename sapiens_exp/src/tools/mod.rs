
//! Bake the acceptance test for the task in the tools
//! Scenario -> (Recipe, Config)
//!
//! Recipe -> (Task, Acceptance criteria)
//!
//! Acceptance criteria -> Tools
//!
//! run_until_completion: (Task, Config, Tools) -> (Trace, Stats, Acceptance
//! results)
//!
//! evaluate::Trial::analyze: (Trace, Stats, Acceptance results) -> Analysis
//!
//! evaluate::Trial::build: (Trace, State, Acceptance results, Task, Analysis)
//! -> Trial

use std::marker::PhantomData;
use std::ops::DerefMut;
use std::sync::Arc;

use sapiens::tools::{Describe, ProtoToolDescribe, ProtoToolInvoke, ToolDescription, ToolUseError};
use serde::{Deserialize, Serialize};
use tokio::sync::Mutex;

/// Scenario 0: Preparing a cereal bowl
pub mod scenario_0;

mod scenario_1;

/// The state of a scenario
pub trait State: Send + Sync {
    /// Reset the state to its initial state
    fn reset(&mut self);
