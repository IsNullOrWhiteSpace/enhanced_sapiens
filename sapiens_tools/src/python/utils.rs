
use pyo3::types::{IntoPyDict, PyDict, PyFloat, PyList, PyTuple};
use pyo3::{PyObject, Python, ToPyObject};
use sapiens::tools::{FieldFormat, ToolDescription};
use serde::{Deserialize, Serialize};
use serde_yaml::Value;

#[derive(thiserror::Error, Debug)]
pub(crate) enum PyConversionError {
    #[error("Invalid conversion: {error}")]
    InvalidConversion { error: String },
    #[error("dict key not serializable: {typename}")]
    DictKeyNotSerializable { typename: String },
    #[error("Invalid cast: {typename}")]
    InvalidCast { typename: String },
}

// inspired from https://github.com/mozilla-services/python-canonicaljson-rs/blob/62599b246055a1c8a78e5777acdfe0fd594be3d8/src/lib.rs#L87-L167
#[allow(clippy::redundant_closure_call)]
pub(crate) fn to_yaml(py: Python, obj: &PyObject) -> Result<Value, PyConversionError> {
    macro_rules! return_cast {
        ($t:ty, $f:expr) => {
            if let Ok(val) = obj.downcast::<$t>(py) {
                return $f(val);
            }
        };
    }

    macro_rules! return_to_value {
        ($t:ty) => {
            if let Ok(val) = obj.extract::<$t>(py) {
                return serde_yaml::to_value(val).map_err(|error| {
                    PyConversionError::InvalidConversion {
                        error: format!("{}", error),
                    }
                });
            }
        };
    }

    if obj.is_none(py) {
        return Ok(Value::Null);
    }

    return_to_value!(String);
    return_to_value!(bool);
    return_to_value!(u64);
    return_to_value!(i64);

    return_cast!(PyDict, |x: &PyDict| {
        let mut map = serde_yaml::Mapping::new();
        for (key_obj, value) in x.iter() {
            let key = if key_obj.is_none() {
                Ok("null".to_string())
            } else if let Ok(val) = key_obj.extract::<bool>() {
                Ok(if val {
                    "true".to_string()
                } else {
                    "false".to_string()
                })
            } else if let Ok(val) = key_obj.str() {
                Ok(val.to_string())
            } else {
                Err(PyConversionError::DictKeyNotSerializable {
                    typename: key_obj
                        .to_object(py)
                        .as_ref(py)
                        .get_type()
                        .name()
                        .map(|x| x.to_string())
                        .unwrap_or_else(|_| "unknown".to_string()),
                })
            };
            map.insert(Value::String(key?), to_yaml(py, &value.to_object(py))?);
        }
        Ok(Value::Mapping(map))
    });

    return_cast!(PyList, |x: &PyList| {
        let v = x
            .iter()
            .map(|x| to_yaml(py, &x.to_object(py)))
            .collect::<Result<Vec<_>, _>>()?;
        Ok(Value::Sequence(v))
    });

    return_cast!(PyTuple, |x: &PyTuple| {
        let v = x
            .iter()
            .map(|x| to_yaml(py, &x.to_object(py)))
            .collect::<Result<Vec<_>, _>>()?;
        Ok(Value::Sequence(v))
    });

    return_cast!(PyFloat, |x: &PyFloat| {
        Ok(Value::Number(serde_yaml::Number::from(x.value())))
    });

    // At this point we can't cast it, set up the error object
    Err(PyConversionError::InvalidCast {
        typename: obj
            .as_ref(py)
            .get_type()
            .name()
            .map(|x| x.to_string())
            .unwrap_or_else(|_| "unknown".to_string()),
    })
}

pub(crate) fn value_to_object(val: Value, py: Python<'_>) -> PyObject {
    match val {
        Value::Null => py.None(),
        Value::Bool(x) => x.to_object(py),
        Value::Number(x) => {
            let oi64 = x.as_i64().map(|i| i.to_object(py));
            let ou64 = x.as_u64().map(|i| i.to_object(py));
            let of64 = x.as_f64().map(|i| i.to_object(py));
            oi64.or(ou64).or(of64).expect("number too large")
        }
        Value::String(x) => x.to_object(py),
        Value::Sequence(x) => {
            let inner: Vec<_> = x.into_iter().map(|x| value_to_object(x, py)).collect();
            inner.to_object(py)
        }
        Value::Mapping(x) => {
            let iter = x
                .into_iter()
                .map(|(k, v)| (value_to_object(k, py), value_to_object(v, py)));
            IntoPyDict::into_py_dict(iter, py).into()
        }
        Value::Tagged(_) => panic!("tagged values are not supported"),
    }
}

/// Format of a field in a tool description
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimpleFormat {
    /// Name of the field
    pub name: String,
    /// Type of the field
    pub r#type: String,
    /// Whether the field is optional
    pub optional: bool,
    /// Description of the field
    pub description: String,
}

impl From<FieldFormat> for SimpleFormat {
    fn from(part: FieldFormat) -> Self {
        SimpleFormat {
            name: part.name,
            r#type: part.r#type,
            optional: part.optional,
            description: part.description,
        }
    }
}

impl ToPyObject for SimpleFormat {
    fn to_object(&self, py: Python<'_>) -> PyObject {
        let dict = PyDict::new(py);
        dict.set_item("name", self.name.to_object(py)).unwrap();
        dict.set_item("type", self.r#type.to_object(py)).unwrap();
        dict.set_item("optional", self.optional.to_object(py))
            .unwrap();
        dict.set_item("description", self.description.to_object(py))
            .unwrap();
        dict.into()
    }
}

/// A simplified version of the tool description
#[derive(Debug, Clone, Serialize, Deserialize)]
pub(crate) struct SimpleToolDescription {
    /// Name of the tool
    pub name: String,
    /// Description of the tool
    pub description: String,
    /// Input format of the tool
    pub parameters: Vec<SimpleFormat>,
    /// Output format of the tool
    pub responses_content: Vec<SimpleFormat>,
}

impl From<ToolDescription> for SimpleToolDescription {
    fn from(desc: ToolDescription) -> Self {
        let parameters = desc
            .parameters
            .fields
            .into_iter()
            .map(|x| x.into())
            .collect();
        let responses_content = desc
            .responses_content
            .fields
            .into_iter()
            .map(|x| x.into())
            .collect();

        SimpleToolDescription {
            name: desc.name,
            description: desc.description,
            parameters,
            responses_content,
        }
    }
}

impl ToPyObject for SimpleToolDescription {
    fn to_object(&self, py: Python<'_>) -> PyObject {
        let dict = PyDict::new(py);
        dict.set_item("name", self.name.clone()).unwrap();
        dict.set_item("description", self.description.clone())
            .unwrap();
        dict.set_item("parameters", self.parameters.clone())
            .unwrap();
        dict.set_item("responses_content", self.responses_content.clone())
            .unwrap();
        dict.into()
    }
}