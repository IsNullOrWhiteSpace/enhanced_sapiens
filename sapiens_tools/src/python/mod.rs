
use std::cmp::Ordering;
use std::collections::HashMap;

use convert_case::{Case, Casing};
use pyo3::indoc::{formatdoc, indoc};
use pyo3::prelude::*;
use pyo3::types::{IntoPyDict, PyDict};
use sapiens::tools::toolbox::{invoke_simple_from_toolbox, Toolbox};
use sapiens::tools::{
    AdvancedTool, Describe, ProtoToolDescribe, ProtoToolInvoke, ToolDescription, ToolUseError,
};
use sapiens_derive::{Describe, ProtoToolDescribe};
use serde::{Deserialize, Serialize};
use serde_yaml::Value;
use tracing::trace;

/// Conversion tools
pub(crate) mod utils;

use crate::python::utils::SimpleToolDescription;

const MAX_OUTPUT_SIZE: usize = 512;

// FUTURE(ssoudan) install pySWIP

/// A tool that runs sandboxed Python code. Use this to transform data.
///
/// - To use another Tool with parameters `input_field_1` and `input_field_2`
///   and result fields `output_field_1` and `output_field_2` use:
/// ```python
/// result = tools.ToolName(input_field_1=..., input_field_2=...)
/// print(result['output_field_1'])
/// print(result['output_field_2'])
/// ```
/// - Only stdout and stderr are captured and made available (limited to 512B
///   total). If the output is larger, use `tools.Conclude` directly from the
///   code.
/// - List available tools with `tools.list()`. And returns a list of
///   `{'name':.., 'description':.., 'parameters':.., 'responses_content':..,
///   }`.
/// - `open`|`exec` are forbidden.
/// - Limited libraries available: urllib3, requests, sympy, numpy,
/// BeautifulSoup4, feedparser, arxiv.
/// - No PIP.
#[derive(Debug, Default, ProtoToolDescribe)]
#[tool(
    name = "SandboxedPython",
    input = "PythonToolInput",
    output = "PythonToolOutput"
)]
pub struct PythonTool {}

/// The input of the Python tool
#[derive(Debug, Serialize, Deserialize, Describe)]
pub struct PythonToolInput {
    /// The Python code to run. MANDATORY
    pub code: String,
}

/// The output of the Python tool
#[derive(Serialize, Deserialize, Describe)]
pub struct PythonToolOutput {
    /// The stdout output of the Python code.
    pub stdout: String,
    /// The stderr output of the Python code.
    pub stderr: String,
}

#[pyclass]
#[derive(Default)]
struct Logging {
    output: String,
}

#[pymethods]
impl Logging {
    fn write(&mut self, data: &str) {
        self.output.push_str(data);
    }
}

#[pyclass(unsendable)]
struct ToolsWrapper {
    toolbox: Toolbox,
    tool_list: Vec<SimpleToolDescription>,
}

impl ToolsWrapper {
    async fn new(toolbox: Toolbox) -> Self {
        let tools = toolbox.describe().await;
        let tool_list = tools
            .into_values()
            .map(SimpleToolDescription::from)
            .collect::<Vec<_>>();

        ToolsWrapper { toolbox, tool_list }
    }
}

#[pymethods]
impl ToolsWrapper {
    // list all tools
    #[pyo3(signature = ())]
    fn list(&self, py: Python<'_>) -> PyResult<PyObject> {
        let tools = self.tool_list.to_object(py);
        Ok(tools)
    }

    // invoke a tool
    #[pyo3(signature = (tool_name, input))]
    fn invoke(
        &self,
        py: Python<'_>,
        tool_name: &str,
        input: Option<&PyDict>,
    ) -> PyResult<PyObject> {
        // convert PyDict to a serde_yaml::Value
        let input = if let Some(input) = input {
            let input: PyObject = input.into();

            utils::to_yaml(py, &input).map_err(|e| {
                pyo3::exceptions::PyException::new_err(format!("Invalid input: {}", e))
            })?
        } else {
            Value::default()
        };

        let (tx, mut rx) = tokio::sync::oneshot::channel::<Result<Value, ToolUseError>>();

        // release the GIL to allow the thread to run
        py.allow_threads(move || {