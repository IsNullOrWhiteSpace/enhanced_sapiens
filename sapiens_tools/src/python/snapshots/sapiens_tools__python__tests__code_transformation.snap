---
source: sapiens_tools/src/python/mod.rs
expression: code
---
class Tools:
    """Wrapper for the tools."""
    def __init__(self, toolbox):
        self.toolbox = toolbox
    def conclude(self, conclusion, original_question):
        """    A tool to conclude a task.
        You have to use this to once you have the answer to the task with your
        conclusion.
        Args:
            conclusion: <str> The final answer for this task. Plain text ONLY. No string interpolation
                supported. MANDATORY. Call directly from `SandboxPython` Tool for long
                answers.
            original_question: <str> The original question that was asked to the user. No string
                interpolation supported, only plain text. MANDATORY.
        """
        return self.toolbox.invoke("Conclude", {"conclusion": conclusion, "original_question": original_question})
    def Conclude(self, conclusion, original_question):
        """    A tool to conclude a task.
        You have to use this to once you have the answer to the task with your
        conclusion.
        Args:
            conclusion: <str> The final answer for this task. Plain text ONLY. No string interpolation
                supported. MANDATORY. Call directly from `SandboxPython` Tool for long
                answers.
            original_question: <str> The original question that was asked to the user. No string
                interpolation supported, only plain text. MANDATORY.
        """
        return self.toolbox.invoke("Conclude", {"conclusion": conclusion, "original_question": original_question})
    def list(self):
        """List the tools."""
        return self.toolbox.list()
tools = Tools(to